pgxn/neon/libpagestore.c

```diff
@@ -64,87 +55,180 @@ int			flush_every_n_requests = 8;
+	 */
+	WaitEventSet    *wes;
+} PageServer;
+
+static PageServer page_servers[MAX_SHARDS];
+static shardno_t  max_attached_shard_no;
+
+static void
+psm_shmem_startup(void)
+{
+	bool found;
+	if (prev_shmem_startup_hook)
+	{
+		prev_shmem_startup_hook();
+	}
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+
+	shard_map = (ShardMap*)ShmemInitStruct("shard_map", sizeof(ShardMap), &found);
+	if (!found)
+	{
+		shard_map->n_shards = 0;
+		pg_atomic_init_u64(&shard_map->begin_update_counter, 0);
+		pg_atomic_init_u64(&shard_map->end_update_counter, 0);
+		AssignPageserverConnstring(page_server_connstring, NULL);
+	}
+	LWLockRelease(AddinShmemInitLock);
 }
 
-static bool
-CheckPageserverConnstring(char **newval, void **extra, GucSource source)
+static void
+psm_shmem_request(void)
 {
-    return strlen(*newval) < MAX_PAGESERVER_CONNSTRING_SIZE;
+#if PG_VERSION_NUM>=150000
+	if (prev_shmem_request_hook)
+		prev_shmem_request_hook();
+#endif
+
+	RequestAddinShmemSpace(sizeof(ShardMap));
 }
 
 static void
-AssignPageserverConnstring(const char *newval, void *extra)
+psm_init(void)
 {
-    if(!PagestoreShmemIsValid())
-        return;
-    LWLockAcquire(pagestore_shared->lock, LW_EXCLUSIVE);
-    strlcpy(pagestore_shared->pageserver_connstring, newval, MAX_PAGESERVER_CONNSTRING_SIZE);
-    pg_atomic_fetch_add_u64(&pagestore_shared->update_counter, 1);
-    LWLockRelease(pagestore_shared->lock);
+	prev_shmem_startup_hook = shmem_startup_hook;
+	shmem_startup_hook = psm_shmem_startup;
+#if PG_VERSION_NUM>=150000
+	prev_shmem_request_hook = shmem_request_hook;
+	shmem_request_hook = psm_shmem_request;
+#else
+	psm_shmem_request();
+#endif
 }
 
-static bool
-CheckConnstringUpdated()
+/*
+ * Reload page map if needed and return number of shards and connection string for the specified shard
+ */
+static shardno_t
+load_shard_map(shardno_t shard_no, char* connstr)
 {
-    if(!PagestoreShmemIsValid())
-        return false;
-    return pagestore_local_counter < pg_atomic_read_u64(&pagestore_shared->update_counter);
+	shardno_t n_shards;
+	uint64 begin_update_counter;
+	uint64 end_update_counter;
+
+	/*
+	 * There is race condition here between backendc and postmaster which can update shard map.
+	 * We recheck update couner after copying connection string to check that configuration was not changed.
+	 */
+	do
+	{
+		begin_update_counter = pg_atomic_read_u64(&shard_map->begin_update_counter);
+		end_update_counter = pg_atomic_read_u64(&shard_map->end_update_counter);
+
+		n_shards = shard_map->n_shards;
+		if (shard_no >= n_shards)
+			elog(ERROR, "Shard %d is greater or equal than number of shards %d", shard_no, n_shards);
+
+		if (connstr)
+			strncpy(connstr, shard_map->shard_connstr[shard_no], MAX_PS_CONNSTR_LEN);
+
+	}
+	while (begin_update_counter != end_update_counter
+		   || begin_update_counter != pg_atomic_read_u64(&shard_map->begin_update_counter)
+		   || end_update_counter != pg_atomic_read_u64(&shard_map->end_update_counter));
```

Comment:
This has the following bug: 
1. Postmaster increments begin_counter
2. Backend reads begin_counter
3. Postmaster starts writing
4. Backend reads end_counter
5. Backend starts reading
6. Postmaster finishes writing and increments end_counter
7. Backend finishes copying
8. Backend performs this check: it passes and moves along with garbage written in its shard map

Reply:
Initial state: begin_counter=update_counter=1
1. Postmaster increments begin_counter
`PM.begin_counter=2`
2. Backend reads begin_counter
`BE.begin_counter=2`
3. Postmaster starts writing
4. Backend reads end_counter
`BE.end_counter=1`
5. Backend starts reading
6. Postmaster finishes writing and increments end_counter
`PM.end_counter=2`
7. Backend finishes copying
8. Backend performs this check
`BE.begin_counter != BE.end_counter` so check is not passed and backend goes to the next iteration perming yet another attempt to read values.

Reply:
Let's try to prove it without TLA+
So we have the following events:
1. PW.write(begin_counter)
2. PM.write(end_counter)
3. PM.begin_write(content)
4. PM.end_write(content)
5. BE.read1(begin_counter)
6. BE.read1(end_counter)
7. BE.begin_read(content)
8. BE.end_read(content)
9. BE.read2(begin_counter)
10. BE.read2(end_content)

There is race condition if (1)
```
BE.begin_read(content) < PM.end_write(content) and BE.end_read(content) > PM.begin_write(content)
```
We have the following statements for event's timing (2):
```
BE.read1(begin_counter) < BE.read1(end_counter) < BE.begin_read(content) < BE.end_read(content) < BE.read2(begin_counter) < BE.read2(end_counter)
```
and (3):
```
PM.write(begin_counter) < PM.begin_write(content) < PM.end_write(content) < PM.write(end_counter)
``` 

By substituting 2) and 3) in 1) we get:

```
BE.read1(end_counter) < PM.write(end_counter)
BE.read2(end_counter) > PM.write(begin_counter)  
BE.read2(begin_counter) > PM.write(begin_counter)  
```

It means that we have just 3 possible schedules for read/write counter events (begin_counter=bc, end_counter=ec):
```
1.  PM.write(bc), BE.read1(ec), BE.read2(bc), BE.read2(ec), PM.write(ec)
2. PM.write(bc), BE.read1(ec), PM.write(ec) , BE.read2(ec)
3. BE.read1(bc), BE.read1(ec),  PM.write(bc) , BE.read2(bc), BE.read2(ec), PM.write(ec) 
```

In 1) `*BE.read1(ec) !=*BE.read2(bc)`
In 2) `*BE.read1(ec) != *BE.read2(ec)`
In 3)  `BE.read1(bc) !=  *BE.read2(bc)`

Reply:
Ah you are right! I had a bug in my spec where the string was not necessarily reloaded once check failed. I fixed that and now the algorithm seems to pass model checking!! Nice